Add extra fields for registration and fix exception when email address exists

Peterede opened the pull request:
I have added some extra fields to the registration form to collect more information about the user, these can be turned on/off in the admin settings.  I also fixed an exception that occurs if the email address is already in the database and gets added twice.

pellaeon commented on service/registrationservice.php:
```diff
@@ -158,6 +163,9 @@ public function validateEmail($email) {
 			return $this->registrationMapper->find($email);//if not found DB will throw a exception
 		} catch (DoesNotExistException $e) {}
+		catch (MultipleObjectsReturnedException $e) {
+			$this->registrationMapper->deleteAll($email);
```
Automatically deleting all objects with same email upon finding multiple entries doesn't seem to be a good idea, since it shouldn't happen in the first place. My original intent was to let user report the error and let the administrator investigate. But I also see the importance of usability here, so please just add a single line to log the error as well.

pellaeon commented:
Also, please fix the broken existing tests because of your changes.

Peterede commented:
When I try and run the tests I get the following error:
~/work/registration$ vendor/phpunit/phpunit/phpunit -c tests/phpunit.xml 
PHP Warning:  require_once(/home/petere/work/registration/tests/../../../tests/bootstrap.php): failed to open stream: No such file or directory in /home/petere/work/registration/tests/autoloader.php on line 12
PHP Fatal error:  require_once(): Failed opening required '/home/petere/work/registration/tests/../../../tests/bootstrap.php' (include_path='.:/usr/share/php') in /home/petere/work/registration/tests/autoloader.php on line 12

Peterede commented:
I have fixed another issue where the validation of the email fails because the user is already registered, the login link shows the html as text instead of a link to go to the login page.

pellaeon commented:
@Peterede Seems that you don't have a full Nextcloud instance for testing (currently it is required). In order to run the test, you need to run phpunit from the `registration` directory that resides in a working Nextcloud instance's `apps/` directory.
The tests will modify the database but will revert any changes in the clean up phase, so I suggest you still use a test Nextcloud instance.

pellaeon commented:
Could you also fix the styles of the create account form so that it looks the same as before? I would prefer the fields are not separated.
Before:
![2018-09-30 3 57 31](https://user-images.githubusercontent.com/916913/46254950-f0124580-c4c9-11e8-9336-a91dd62221c4.png)

After:
![2018-09-30 3 58 06](https://user-images.githubusercontent.com/916913/46254956-01f3e880-c4ca-11e8-887b-89aa32461a05.png)
![2018-09-30 3 58 35](https://user-images.githubusercontent.com/916913/46254957-01f3e880-c4ca-11e8-8436-5fd87bd375b6.png)

- [ ] Continuous input fields
- [ ] Icons are shifted (the person and eye icon)
- [ ] Add icons for new fields so that the design looks consistent

Peterede commented:
I have fixed the issues except for the icons on the new fields, I will work on this

Peterede commented:
All changes have been made now

pellaeon commented:
Sorry, but there are a few more problems that I found out:
1. please use meaningful icons for each field, instead just putting `contacts-dark.svg` everywhere
2. `contacts-dark.svg` doesn't exist in Nextcloud's master, this will cause error when rendering the form, please choose the icons that exist in both latest Nextcloud and ownCloud
3. the timezone selection is empty on Nextcloud master
4. please move the fields that have a title (country, language, timezone) to the bottom, because they cannot be combined with other fields. (keep all combined fields together at the top)

Peterede commented:
1. For some fields there are no meaningful icons available, the nextcloud personal settings uses this icon for multiple fields so I used that as reference.  
2. contacts-dark.svg is in /core/img/places/contacts-dark.svg on nextcloud master
3. I have added the moment-timezone-with-data.js file to the app
4. I have moved the lists to the bottom so that the input fields are grouped

pellaeon commented:
2. I'm sorry, but I can't find it there: https://github.com/nextcloud/server/tree/master/core/img/places
Maybe it's related to theming, do you have theming enabled? I don't. https://github.com/nextcloud/server/issues/10443

pellaeon commented:
Also, `moment-timezone-with-data.js` is a 190KB file, could you make it load **only** when the timezone field is enabled?

Peterede commented:
In my checkout of master the file is located in core/img/places but if I download the source from github or view it in the web browser the file is not there.  The latest commit shows the file was deleted https://github.com/nextcloud/server/commit/29ff7efe9a5be16b133a1ee4e43d6d2155b6a21c
I replaced contacts-dark.svg with social.svg which is the same icon

Peterede commented on controller/registercontroller.php:
```diff
@@ -141,18 +160,38 @@ public function verifyToken($token) {
+
 		try {
-			$user = $this->registrationService->createAccount($registration, $username, $password);
+			$user = $this->registrationService->createAccount($registration, $username, $password, $fullname, $country, $language, $phoneno, $timezone, $company);
```
I have made the changes

thoka commented:
I would find this patch very usefull. What has to be done to merge it ?

nickvergessen commented:
> I would find this patch very usefull. What has to be done to merge it ?

Due to the number of conflicts I guess it has to be redone
